tests/test_log_vm.py

```diff
@@ -29,18 +32,27 @@ def test_redis_service_running(self):
         self.assertTrue(self._service_is_active("redis"))
 
     def test_logs_are_flowing(self):
-        cmd_output = self._run("ls -1 /home/user/QubesIncomingLogs")
-        log_dirs = cmd_output.split("\n")
-        # Confirm AppVMs are sending logs
-        self.assertTrue("sd-app" in log_dirs)
-        # The following will only have logs if the machine has booted,
-        # which is not guaranteed given that we randomize test order.
-        # self.assertTrue("sd-devices" in log_dirs)
-        # self.assertTrue("sd-proxy" in log_dirs)
-        # self.assertTrue("sd-viewer" in log_dirs)
+        """
+        To test that logs work, we run a unique command in each VM we care
+        about that gets logged, and then check for that string in the logs.
+        """
+        # Random string, to avoid collisions with other test runs
+        token = "".join(secrets.choice(string.ascii_uppercase) for _ in range(10))
+
+        # TODO: test a sd-viewer based dispVM
+        for vm_name in ["sd-app", "sd-devices", "sd-proxy", "sd-gpg"]:
+            # The sudo call will make it into syslog
+            subprocess.check_call(["qvm-run", vm_name, f"sudo echo {token}"])
+
+        # Now look in the log files to find the token
+        for vm_name in ["sd-app", "sd-devices", "sd-proxy"]:
+            contents = self._get_file_contents(f"/home/user/QubesIncomingLogs/{vm_name}/syslog.log")
+            self.assertIn(token, contents)
+
+        # But nothing for sd-gpg
+        self.assertFalse(self._fileExists("/home/user/QubesIncomingLogs/sd-gpg/syslog.log"))
```

Comment:
Do you see any point in parameterizing this now, such that either:
1. we have sets `logs_expected_from` and `logs_not_expected_from` that we enforce; or even
2. we have a set `log_expected_from` that we check strictly against all the VMs on the system?

(2) might well be overkill, but these are the two possibilities that occurred to me.

Reply:
Yeah, I think that's a good idea, let me see...

Reply:
Done! There's one todo about sd-viewer, because we can't actually run/start sd-viewer, we have to use a disposable based on it. There's a workaround for it in the sd-viewer tests, but it's not reusable yet. Once we're pytest and can have easily shared fixtures, it should be trivial.